### ISSUE TITLE
Does the size of batch-size affect the training results?

### ISSUE THREAD

**Issue by youmi-zym** [ClementPinard/SfmLearner-Pytorch#32]
Hi,
I have run the train.py with the command blow on KITTI-raw-data :
`python3 train.py /path/to/the/formatted/data/ -b4 -m0 -s2.0 --epoch-size 1000 --sequence-length 5 --log-output --with-gt`
Otherwise the batch-size=80, and the train(41664)/vaild(2452) split is different.
The result I get is:
disp:
Results with scale factor determined by GT/prediction ratio (like the original paper) : 
`
abs_rel,     sq_rel,        rms,    log_rms,         a1,         a2,         a3
    0.2058,     1.6333,     6.7410,     0.2895,     0.6762,     0.8853,     0.9532

pose:
Results 10
	        ATE,         RE
mean 	     0.0223,     0.0053
std 	     0.0188,     0.0036

Results 09
	        ATE,         RE
mean 	     0.0284,     0.0055
std 	     0.0241,     0.0035
`
You can see that there's still a quiet big margin with yours:
Abs Rel | Sq Rel | RMSE | RMSE(log) | Acc.1 | Acc.2 | Acc.3
0.181 | 1.341 | 6.236 | 0.262 | 0.733 | 0.901 | 0.964

I think there is no other factors causing this difference, otherwise the batch-size and data split. Therefore, does the size of batch-size affect the training results?

What's more, when I try to train my model with two Titan GPUs, batch-size=80*2=160, the memory usage of each GPU is:
GPU0: about 11G,    GPU1: about 6G.
There is a huge memory usage difference between two GPUs, and it seriously impacts multi-gpu trianing. 
And then I find the loss calculations are all placed on the first GPU, actually the memory is mainly used to calculate the 4 scales of depth photometric_reconstruction_loss, and we can just move  some scales to the cuda:0, and others to cuda:1, it might be better I think.

**Comment by anuragranj** [comment]
@youmi-zym By the way, I followed this for multigpu. https://github.com/NVIDIA/flownet2-pytorch
Also, you don't want to change the batch size, even if you use multigpu.

**Comment by youmi-zym** [comment]
@ClementPinard  @anuragranj Yeah, thanks for your help, and I have implemented multigpu loss function with your advise. However, I still have some experiment results to share with you ( [smooth-loss](https://github.com/ClementPinard/unsupervised-depthnet/blob/master/loss_functions.py#L55), [texture-smooth-loss](https://github.com/ClementPinard/unsupervised-depthnet/blob/master/loss_functions.py#L100) ) :
![image](https://user-images.githubusercontent.com/17737537/45940767-78e23a80-c00d-11e8-9dcc-c5ce303b5231.png)

It means these experiments run with the top command, the smooth-loss, texture-smooth-loss are mentioned upon, and the weight is set as the command. It shows that the batch-size exactly makes modest influence to the result, but when visualize the result, I find out that small batch-size is good for close object, but big batch-size pays more attention to optimize the far away objects. 
The point is that I still get worse result than yours... It's wired.....

**Comment by ClementPinard** [comment]
According to the original author, quality worsens after 160K iterations anyway (https://github.com/tinghuiz/SfMLearner/issues/42)
Other than that, That's odd that  `batch size 4 and 1 GPU` gives you worse results than the ones reported in the README.
I'll try a training, and see if there is any regression.

How did you split your training dataset ?

**Comment by ClementPinard** [comment]
Here are my results with my own split : 
```
Results with scale factor determined by GT/prediction ratio (like the original paper) : 
   abs_rel,     sq_rel,        rms,    log_rms,         a1,         a2,         a3
    0.1840,     1.3708,     6.3823,     0.2669,     0.7187,     0.9023,     0.9627
```
I used a regular smooth loss and a trained on a clean version of the repo.
What pretrained model did you use ? Make sure you used the "model_best" version and not the "checkpoint"

I'll check with your split

**Comment by ClementPinard** [comment]
Results with your split, using model_best : 
```
Results with scale factor determined by GT/prediction ratio (like the original paper) : 
   abs_rel,     sq_rel,        rms,    log_rms,         a1,         a2,         a3
    0.1854,     1.3986,     6.4104,     0.2687,     0.7149,     0.8985,     0.9619
```

Results with your split, using checkpoint :
```
Results with scale factor determined by GT/prediction ratio (like the original paper) : 
   abs_rel,     sq_rel,        rms,    log_rms,         a1,         a2,         a3
    0.2040,     1.8203,     6.6266,     0.2914,     0.6971,     0.8848,     0.9510
```

As such, I think you only used the `checkpoint.pth.tar` . This is consistent with author's claim that you eventually end up with worse results if you keep on training after more than 140K iterations.

**Comment by youmi-zym** [comment]
@ClementPinard Thanks for your work and reply, I use the "model_best" version... I will try again with 140k iterations.  Thanks a lot.

**Comment by ClementPinard** [comment]
Here is my exact dataset, in case there has been some regression in the data preporcessing script :

https://mega.nz/#!OIEwEQ4a!Yz5aRFjPHxNwCV2sxIslgWPfppAj_WOpthNTqWUvByo

**Comment by youmi-zym** [comment]
Thanks a lot ! I will check it and try again !

**Comment by youmi-zym** [comment]
The result is still bad....
First I clone the latest repo, use the KITTI-GT you give me, and train with the command below:
`python3 train.py /home/data/UnsupervisedDepth/KITTI-raw/KITTI_GT -b4 -m0 -s2.0 --epochs 140 --epoch-size 1000 --sequence-length 5 --log-output --with-gt`
Then, I use the command below to test:
`python3 test_disp.py 	--pretrained-dispnet /home/data/youmi/projects/depth/SfmLearner-Pytorch/checkpoints/KITTI_GT,140epochs,epoch_size1000,seq5,s2.0/10-13-10:32/dispnet_model_best.pth.tar \
			--dataset-dir /home/data/UnsupervisedDepth/KITTI-raw/raw_data_KITTI \
			--dataset-list ./kitti_eval/test_files_eigen.txt`

Finally, the result I get:

no PoseNet specified, scale_factor will be determined by median ratio, which is kiiinda cheating            (but consistent with original paper)
getting test metadata ... 
100%|████████████████████████████████████████| 697/697 [00:04<00:00, 156.10it/s]
697 files to test
100%|█████████████████████████████████████████| 697/697 [00:46<00:00, 14.91it/s]
Results with scale factor determined by GT/prediction ratio (like the original paper) : 
   abs_rel,     sq_rel,        rms,    log_rms,         a1,         a2,         a3
    0.1923,     1.4461,     6.4234,     0.2705,     0.7131,     0.8977,     0.9608

**Comment by ClementPinard** [comment]
Thanks for your insight !

Since it got worse with more training, there may be a problem with the validation set since it is supposed to keep the best network on it. Maybe it's not representative enough compared to the test set ?

Anyway, this auto supervision problem is very hard to make converge correctly, and to make interpretation on the different results.

Good luck for you research!

**Comment by youmi-zym** [comment]
@ClementPinard Thanks very much
